src/main/java/uk/gov/companieshouse/api/accounts/exception/handler/GlobalExceptionHandler.java

```diff
@@ -0,0 +1,93 @@
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.context.request.WebRequest;
+import org.springframework.web.server.ResponseStatusException;
+import org.springframework.web.servlet.NoHandlerFoundException;
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+import uk.gov.companieshouse.api.accounts.exception.ApiException;
+import uk.gov.companieshouse.logging.Logger;
+import uk.gov.companieshouse.logging.LoggerFactory;
+
+@ControllerAdvice
+public class GlobalExceptionHandler {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(APPLICATION_NAME_SPACE);
+
+  @ExceptionHandler(value = {DataAccessException.class, IOException.class,
+      IllegalArgumentException.class, IllegalStateException.class, NullPointerException.class,
+      RuntimeException.class, NoSuchAlgorithmException.class, Exception.class})
+  @ResponseStatus(value = HttpStatus.INTERNAL_SERVER_ERROR)
+  protected void handleException(Exception ex) {
+    logError(ex, getExceptionMessage(ex));
+  }
+
+  @ExceptionHandler(value = {NoHandlerFoundException.class})
+  @ResponseStatus(value = HttpStatus.NOT_FOUND)
+  protected void handleNoHandlerFoundException(Exception ex) {
+    logError(ex, ExceptionMessage.NO_HANDLER_FOUND_EXCEPTION);
+  }
+
+  @ExceptionHandler(value = {ApiException.class})
+  protected void handleApiException(ApiException apiex) {
+    logError(apiex, apiex.exceptionMessage);
+    throw new ResponseStatusException(getHttpStatus(apiex), apiex.exceptionMessage.getMessage());
+  }
+
+
+  private void logError(Exception ex, ExceptionMessage exceptionMessage) {
+    HashMap<String, Object> message = new HashMap<>();
+    message.put("message", ex.getMessage());
+    message.put("error", exceptionMessage.getError());
+    LOGGER.error(exceptionMessage.getMessage(), ex, message);
+  }
+
+  private ExceptionMessage getExceptionMessage(Exception ex) {
+    if (ex instanceof DataAccessException) {
+      return ExceptionMessage.DATA_ACCESS_EXCEPTION;
+    } else if (ex instanceof IOException) {
+      return ExceptionMessage.IO_EXCEPTION;
+    } else if (ex instanceof IllegalArgumentException) {
+      return ExceptionMessage.ILLEGAL_ARGUMENT_EXCEPTION;
+    } else if (ex instanceof IllegalStateException) {
+      return ExceptionMessage.ILLEGAL_STATE_EXCEPTION;
+    } else if (ex instanceof NullPointerException) {
+      return ExceptionMessage.NULL_POINTER_EXCEPTION;
+    } else if (ex instanceof RuntimeException) {
+      return ExceptionMessage.RUN_TIME_EXCEPTION;
+    }
+    return ExceptionMessage.EXCEPTION;
+  }
+
+  private HttpStatus getHttpStatus(ApiException apiEx) {
+    switch (apiEx.exceptionMessage) {
+      case ACCOUNT_LINK_PRESENT:
+        return HttpStatus.CONFLICT;
```

Comment:
Why is an API exception being thrown in the code expecting this handler to pick up the exception and convert to the correct HTTP status? This is more likely to lead to bugs. The correct HTTP status should be returned from the controller where the error condition occurs.